File: JavaScript/Graph - DFS/841. Keys and Rooms.js
/*  841. Keys and Rooms
Difficulty: Medium
Submission Link: https://leetcode.com/problems/keys-and-rooms/submissions/1381349579/ */

var canVisitAllRooms = function (rooms) {

    let vis = new Array(rooms.length), stack = [0], count = 1
    vis[0] = 1
    while (stack.length) {
        let keys = rooms[stack.pop()]
        for (let k of keys)
            if (!vis[k]) stack.push(k), vis[k] = 1, count++
    }
    return rooms.length === count
};

File: Python3/Graph - DFS/841. Keys and Rooms.py
""" 841. Keys and Rooms
Difficulty: Medium
Submission Link: https://leetcode.com/problems/keys-and-rooms/submissions/1381349727/ """

class Solution:
    def canVisitAllRooms(self, rooms: list[list[int]]) -> bool:
        n = len(rooms)
        vis = [False] * n
        stack = [0]
        vis[0] = True
        count = 1

        while stack:
            current_room = stack.pop()
            for key in rooms[current_room]:
                if not vis[key]:
                    stack.append(key)
                    vis[key] = True
                    count += 1
        
        return count == n